File: caracal/sdk/adapters/mock.py
"""
Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
Caracal, a product of Garudex Labs

Mock transport adapter for local testing.
"""

from __future__ import annotations

from typing import Any, Dict, Optional, Tuple

from caracal.sdk.adapters.base import BaseAdapter, SDKRequest, SDKResponse


class MockAdapter(BaseAdapter):
    """In-memory mock adapter for unit tests.

    Args:
        responses: Mapping from ``(method, path)`` tuples to
            ``SDKResponse`` instances.

    Example::

        adapter = MockAdapter({
            ("GET", "/agents"): SDKResponse(status_code=200, body=[]),
        })
    """

    def __init__(
        self,
        responses: Optional[Dict[Tuple[str, str], SDKResponse]] = None,
    ) -> None:
        self._responses: Dict[Tuple[str, str], SDKResponse] = responses or {}
        self._sent: list[SDKRequest] = []

    async def send(self, request: SDKRequest) -> SDKResponse:
        self._sent.append(request)
        key = (request.method.upper(), request.path)
        if key in self._responses:
            return self._responses[key]
        return SDKResponse(
            status_code=404,
            headers={},
            body={"error": "not mocked"},
            elapsed_ms=0.0,
        )

    def close(self) -> None:
        self._responses.clear()
        self._sent.clear()

    @property
    def is_connected(self) -> bool:
        return True

    @property
    def sent_requests(self) -> list[SDKRequest]:
        """All requests that have been sent through this adapter."""
        return list(self._sent)

File: sdk/node-sdk/src/adapters/mock.ts
/**
 * Copyright (C) 2026 Garudex Labs. All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * Mock transport adapter for testing.
 */

import { BaseAdapter, SDKRequest, SDKResponse } from './base';

export class MockAdapter extends BaseAdapter {
  private readonly responses: Map<string, SDKResponse>;
  private readonly sent: SDKRequest[] = [];

  constructor(responses?: Map<string, SDKResponse>) {
    super();
    this.responses = responses ?? new Map();
  }

  /** Add a mocked response keyed by `METHOD /path`. */
  mock(method: string, path: string, response: SDKResponse): void {
    this.responses.set(`${method.toUpperCase()} ${path}`, response);
  }

  async send(request: SDKRequest): Promise<SDKResponse> {
    this.sent.push(request);
    const key = `${request.method.toUpperCase()} ${request.path}`;
    const match = this.responses.get(key);
    if (match) return match;
    return {
      statusCode: 404,
      headers: {},
      body: { error: 'not mocked' },
      elapsedMs: 0,
    };
  }

  close(): void {
    this.responses.clear();
    this.sent.length = 0;
  }

  get isConnected(): boolean {
    return true;
  }

  /** All requests that have been sent through this adapter. */
  get sentRequests(): SDKRequest[] {
    return [...this.sent];
  }
}
